File: backend/app/api/v1/endpoints/foresight.py
"""Planification prospective : projections ML + scénarios + plan IA."""
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session
from app.db.session import get_db
from app.ml.forecasting import forecast_series, project_indicator
from app.services.planning import generate_recommendations
from app.services.ai_planner import generate_plan
from app.services.structured_plan import build_structured_plan

router = APIRouter(prefix="/foresight", tags=["foresight"])


def _territory_context(territory_id: int, db: Session):
    from app.models.territory import Territory, Zone, Building
    from app.models.indicator import Indicator
    t = db.get(Territory, territory_id)
    name = t.name if t else f"Territoire {territory_id}"
    indicators = {i.key: i.value for i in
                  db.query(Indicator).filter(Indicator.territory_id == territory_id).all()}
    zone_ids = [z.id for z in db.query(Zone).filter(Zone.territory_id == territory_id).all()]
    buildings = db.query(Building).filter(Building.zone_id.in_(zone_ids)).all() if zone_ids else []
    bdata = [{"energy_class": b.energy_class, "surface_m2": b.surface_m2} for b in buildings]
    return name, indicators, bdata


@router.get("/{territory_id}/forecast")
def forecast(territory_id: int, target_year: int = 2035):
    """Projection ML de la consommation énergétique nationale."""
    return forecast_series(target_year)


@router.get("/{territory_id}/scenarios")
def scenarios(territory_id: int, db: Session = Depends(get_db), horizon: int = 10):
    """Trois trajectoires prospectives d'un indicateur clé (perf. énergétique)."""
    _, indicators, _ = _territory_context(territory_id, db)
    current = indicators.get("energy_performance", 75)
    return {
        "indicator": "Performance énergétique (%)",
        "current": current,
        "trajectories": {
            "tendanciel": project_indicator(current, 1.0, horizon),
            "volontariste": project_indicator(current, 2.5, horizon),
            "objectif_2030": project_indicator(current, 4.0, horizon),
        },
    }


class PlanRequest(BaseModel):
    horizon: int = 10


@router.post("/{territory_id}/plan")
def plan(territory_id: int, payload: PlanRequest, db: Session = Depends(get_db)):
    """Génère un plan d'action prospectif (ML + diagnostic + IA générative)."""
    name, indicators, bdata = _territory_context(territory_id, db)
    fc = forecast_series(2025 + payload.horizon)
    recs = generate_recommendations(indicators, bdata)
    # Daïras de la wilaya (pour que le plan IA les cite)
    from app.data.admin_divisions import ADMIN_DIVISIONS
    from app.models.territory import Territory as _T
    _t = db.get(_T, territory_id)
    _wc = (_t.wilaya_code or "").zfill(2) if _t else ""
    _wd = ADMIN_DIVISIONS.get(_wc)
    subdivisions = list(_wd["dairas"].keys()) if _wd else []
    result = generate_plan(name, indicators, fc, recs, payload.horizon, subdivisions)
    return {
        "territory": name,
        "horizon": payload.horizon,
        "source": result["source"],
        "forecast_summary": {
            "from": fc["history"][-1],
            "to": fc["forecast"][-1],
            "cagr_pct": fc.get("historical_cagr_pct"),
            "r2": fc["r2"],
        },
        "plan_markdown": result["plan"],
    }


@router.post("/{territory_id}/structured-plan")
def structured_plan(territory_id: int, payload: PlanRequest, db: Session = Depends(get_db)):
    """Plan de planification structuré (phases, actions, budgets, cibles)."""
    name, indicators, bdata = _territory_context(territory_id, db)
    fc = forecast_series(2025 + payload.horizon)
    recs = generate_recommendations(indicators, bdata)
    return build_structured_plan(name, indicators, fc, recs, payload.horizon)


@router.get("/{territory_id}/subdivisions")
def foresight_subdivisions(territory_id: int, db: Session = Depends(get_db)):
    """Liste des daïras/communes de la wilaya + récapitulatif prospectif par niveau."""
    from app.models.territory import Territory
    from app.data.admin_divisions import ADMIN_DIVISIONS
    from app.data.energy_data import energy_distribution, estimate_buildings
    import hashlib

    t = db.get(Territory, territory_id)
    if not t:
        from fastapi import HTTPException
        raise HTTPException(404, "Territoire introuvable")

    wcode = (t.wilaya_code or "").zfill(2)
    wilaya_data = ADMIN_DIVISIONS.get(wcode)

    def proj(seed: str, base_perf: float):
        # projection de performance à horizon (amélioration progressive)
        h = int(hashlib.md5(seed.encode()).hexdigest(), 16)
        current = max(15, min(95, round(base_perf + (h % 11 - 5))))
        target = min(98, current + 8 + (h % 10))  # objectif à ~10 ans
        return current, target

    try:
        prof = territory_profile(territory_id, db)
        base_perf = prof.get("energy_performance", 70)
    except Exception:
        base_perf = 70

    dairas = []
    if wilaya_data:
        pop_per = (t.population or 0) / max(1, sum(len(cs) for cs in wilaya_data["dairas"].values()))
        for dname, communes in wilaya_data["dairas"].items():
            clist = []
            for cname in communes:
                cur, tgt = proj(f"{cname}-proj", base_perf)
                clist.append({"name": cname, "current": cur, "target": tgt, "gain": tgt - cur})
            davg_cur = round(sum(c["current"] for c in clist) / len(clist)) if clist else base_perf
            davg_tgt = round(sum(c["target"] for c in clist) / len(clist)) if clist else base_perf
            dairas.append({"name": dname, "current": davg_cur, "target": davg_tgt,
                          "gain": davg_tgt - davg_cur, "communes": clist})

    return {
        "territory_id": t.id, "territory_name": t.name, "wilaya_code": wcode,
        "dairas": dairas, "has_detail": len(dairas) > 0,
    }

File: frontend/src/api/foresight.ts
import { apiClient } from "./client";

export interface ForecastPoint { year: number; value: number; }
export interface Forecast {
  metric: string;
  history: ForecastPoint[];
  forecast: ForecastPoint[];
  r2: number;
  historical_cagr_pct: number;
  model: string;
}
export interface Scenarios {
  indicator: string;
  current: number;
  trajectories: Record<string, ForecastPoint[]>;
}
export interface PlanResult {
  territory: string;
  horizon: number;
  source: "mistral" | "fallback";
  forecast_summary: { from: ForecastPoint; to: ForecastPoint; cagr_pct: number; r2: number };
  plan_markdown: string;
}

export const fetchForecast = async (territoryId: number, targetYear = 2035): Promise<Forecast> => {
  const { data } = await apiClient.get<Forecast>(`/foresight/${territoryId}/forecast`, {
    params: { target_year: targetYear },
  });
  return data;
};
export const fetchScenarios = async (territoryId: number, horizon = 10): Promise<Scenarios> => {
  const { data } = await apiClient.get<Scenarios>(`/foresight/${territoryId}/scenarios`, {
    params: { horizon },
  });
  return data;
};
export const generatePlan = async (territoryId: number, horizon = 10): Promise<PlanResult> => {
  const { data } = await apiClient.post<PlanResult>(`/foresight/${territoryId}/plan`, { horizon });
  return data;
};

export interface PlanAction {
  name: string; lever: string; budget_m_da: number; deadline: string; priority: string;
}
export interface PlanPhase {
  title: string; period: string; focus: string;
  actions: PlanAction[]; total_budget_m_da: number;
  target: { indicator: string; from: number; to: number };
}
export interface StructuredPlan {
  territory: string; horizon_years: number; period: string;
  summary: {
    current_performance: number; target_performance: number;
    current_resilience: number; renewable_target: number;
    total_budget_m_da: number; energy_forecast_2035_gwh: number | null;
  };
  phases: PlanPhase[];
  key_levers: string[];
}

export const fetchStructuredPlan = async (territoryId: number, horizon = 10): Promise<StructuredPlan> => {
  const { data } = await apiClient.post<StructuredPlan>(`/foresight/${territoryId}/structured-plan`, { horizon });
  return data;
};
